src/main/scala/com/redhat/et/silex/cluster/KMedoids.scala

```diff
@@ -0,0 +1,422 @@
+/*
+ * This file is part of the "silex" library of helpers for Apache Spark.
+ *
+ * Copyright (c) 2015 Red Hat, Inc.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.c
+ */
+
+package com.redhat.et.silex.cluster
+
+import scala.util.Random
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.Logging
+import org.apache.spark.util.random.XORShiftRandom
+
+/** An object for training a K-Medoid clustering model on Seq or RDD data.
+  *
+  * Data is required to have a metric function defined on it, but it does not require an algebra
+  * over data elements, as K-Means clustering does.
+  */
+class KMedoids[T] private (
+  private var metric: (T, T) => Double,
+  private var k: Int,
+  private var maxIterations: Int,
+  private var epsilon: Double,
+  private var fractionEpsilon: Double,
+  private var sampleSize: Int,
+  private var seed: Long) extends Serializable with Logging {
+
+  /** Constructs a KMedoids instance from a given metric function
+    *
+    * @param metric A metric function defined on data elements
+    */
+  def this(metric: (T, T) => Double) = this(
+    metric,
+    KMedoids.default.k,
+    KMedoids.default.maxIterations,
+    KMedoids.default.epsilon,
+    KMedoids.default.fractionEpsilon,
+    KMedoids.default.sampleSize,
+    KMedoids.default.seed)
+
+  private def medoidDist(e: T, mv: Seq[T]) = {
+    val n = mv.length
+    var mMin = Double.MaxValue
+    var j = 0
+    while (j < n) {
+      val m = metric(e, mv(j))
+      if (m < mMin) { mMin = m }
+      j += 1
+    }
+    mMin
+  }
+
+  private def medoidIdx(e: T, mv: Seq[T]) = {
+    val n = mv.length
+    var mMin = Double.MaxValue
+    var jMin = 0
+    var j = 0
+    while (j < n) {
+      val m = metric(e, mv(j))
+      if (m < mMin) {
+        mMin = m
+        jMin = j
+      }
+      j += 1
+    }
+    jMin
+  }
+
+  private def medoidCost(e: T, data: Seq[T]) = data.iterator.map(metric(e, _)).sum
+  private def medoid(data: Seq[T]) = data.iterator.minBy(medoidCost(_, data))
+  private def modelCost(mv: Seq[T], data: Seq[T]) = data.iterator.map(medoidDist(_, mv)).sum
+
+  /** Set the distance metric to use over data elements
+    *
+    * @param metric The distance metric
+    * @return This instance after setting metric
+    */
+  def setMetric(metric: (T, T) => Double): this.type = {
```

Comment:
Adhering to Spark convention, the `run` method generates a clustering using whatever parameters are set at the time and returns a `KMedoidsModel` object, which is a static object containing its own `metric` and `k` information.

So, if you change the metric, it will use the new metric on any subsequent invocations of `run`, as with all the other parameters